File: 1802-NumberOfStudentsUnableToEatLunch/1802-NumberOfStudentsUnableToEatLunch.java
// Last updated: 6/10/2026, 7:19:55 PM
class Solution {
    public int countStudents(int[] students, int[] sandwiches) {
        Deque<Integer> queue = new LinkedList<>();
        int idx = 0;
        int circles = 0 , sqaures = 0;
        for (var i : students) {
            queue.add(i);
            if (i == 0) circles++;
            else sqaures++;
        }
        while (!queue.isEmpty()) {
            if (queue.peek() == sandwiches[idx]) {
                if (sandwiches[idx] == 0) circles--;
                else sqaures--;
                idx++;
                queue.remove();
            } else if (sandwiches[idx] == 0) {
                if (circles == 0) {
                    return sqaures; 
                } else {
                    queue.add(queue.remove());
                }
            } else {
                if (sqaures == 0) return circles;
                else queue.add(queue.remove());
            }
        }
        return 0;
    }   
}

File: 1802-NumberOfStudentsUnableToEatLunch/1802-NumberOfStudentsUnableToEatLunch.cpp
// Last updated: 6/10/2026, 7:20:02 PM
class Solution {
public:
    int countStudents(vector<int>& students, vector<int>& sandwiches) {
        reverse(sandwiches.begin() , sandwiches.end());
        queue<int> q;
        int cnt = 0 , circles = 0 , squares = 0;
        for (auto i : students) {
            q.push(i);
            (i == 0 ? circles : squares)++;
        }
        while (!q.empty()) {
            if (q.front() == sandwiches.back()) {
                (q.front() == 0 ? circles : squares)--;
                q.pop();
                sandwiches.pop_back();
            } else if (sandwiches.back() == 0) {
                if (circles == 0) {
                    cnt += squares;
                    break;
                } else {
                    q.push(q.front());
                    q.pop();
                }  
            } else {
                if (squares == 0) {
                    cnt += circles;
                    break;
                } else {
                    q.push(q.front());
                    q.pop();
                }
            }
        }
        return cnt;
    }
};